src/flexible_type/flexible_type_converter.hpp

```diff
@@ -10,657 +10,899 @@
+template <typename T> static flexible_type convert_to_flexible_type(const T& t);
 
-// forward declaration
-template <typename T>
-struct is_flexible_type_convertible;
+namespace flexible_type_internals {
 
-// forward declaration
-template <typename... Args>
-struct all_flexible_type_convertible;
+////////////////////////////////////////////////////////////////////////////////
 
-/**
- * all_arithmetic<Args...>::value is true if every type in 
- * Args are arithmetic types (numeric).
+/** The conversion path is chosen based on a list of possible
+ *  conversions ranked by priority, with the lowest numbered converter
+ *  that matches the particular type chosen as the conversion path.
+ *  The constants below declare the particular converters available,
+ *  and the actual definitions -- defined as specialized structs --
+ *  are given after that.
+ *
+ *  The conversion mechanism works by testing the matches() constexpr
+ *  condition on each numbered struct, then choosing the struct number
+ *  that matches with no lower numbers matching.  A static_assert
+ *  failure occurs if there are no matches.
  */
-template <typename... Args>
-struct all_arithmetic {
-  typedef boost::mpl::vector<Args...> type_sequence;
-  // makes an mpl::vector<true_, false_ ....> where each element transforms
-  // each arg to whether it is flexible_type convertible
-  typedef typename 
-      boost::mpl::transform<type_sequence, 
-      std::is_arithmetic<boost::mpl::_1>>::type transformed_sequence; 
-
-  // the number of good types (number which are true)
-  typedef typename 
-      boost::mpl::count<transformed_sequence, 
-                        std::integral_constant<bool, true>>::type num_good;
- 
-  // true if all args are convertible.  
-  static constexpr bool value = (num_good::value == sizeof...(Args));
-};
 
+// The obvious case of another flexible type.
+static constexpr int CVTR__FLEXIBLE_TYPE                  = 1;
+
+// Numeric types.
+static constexpr int CVTR__FLOATING_POINT                 = 2;
+static constexpr int CVTR__INTEGER                        = 3;
+
+// Vectors -- sequences of numeric elements that can be exactly
+// converted into a flex_vec type.  Prioritized above lists.
+static constexpr int CVTR__FLEX_VEC_EXACT                 = 4;
+static constexpr int CVTR__FLEX_VEC_CONVERTIBLE_SEQUENCE  = 5;
+static constexpr int CVTR__FLEX_VEC_CONVERTIBLE_PAIR      = 6;
+static constexpr int CVTR__FLEX_VEC_CONVERTIBLE_TUPLE     = 7;
+
+// Dictionaries -- sequences convertible to an exact match.
+static constexpr int CVTR__FLEX_DICT_EXACT                = 8;
+static constexpr int CVTR__FLEX_DICT_CONVERTIBLE_SEQUENCE = 9;
+static constexpr int CVTR__FLEX_DICT_CONVERTIBLE_MAPS     = 10;
+
+// Lists -- any other sequences that cannot be converted into a
+// dictionary or vector but can be converted into a list of flexible
+// type.
+static constexpr int CVTR__FLEX_LIST_EXACT                = 11;
```

Comment:
Can all the EXACT cases be merged? Previously is_flexible_type_member<T>::value is used.

Reply:
Yes, they can be merged.  I can do that easily.  Will do.